### ISSUE TITLE
Recipe import fails (Bad Gateway error)

### ISSUE THREAD

**Issue by tdittmar** [TandoorRecipes/recipes#1256]
## Issue
The Tandoor installation works (I can create recipes, etc.), but I can not import recipes. The site gives a red error box after a couple of seconds saying

```
An error occurred while trying to import this recipe!<html>
 <head><title>502 Bad Gateway</title></head>
 <body>
 <center><h1>502 Bad Gateway</h1></center>
 <hr><center>nginx/1.21.5</center>
 </body>
 </html>
```

## Setup Info
Version: 1.0.4
OS: Ubuntu 18.04.6 LTS

Docker installation performed as recommended. Only change: nginx uses port mapping from local port 23080 to 80 and I have another nginx proxying to that port (configured in Plesk).

### `.env`
Please include your `.env` config file (**make sure to remove/replace all secrets**)
```
DEBUG=0
SQL_DEBUG=0
ALLOWED_HOSTS=*

# ---------------------------- REQUIRED -------------------------
SECRET_KEY=...
# ---------------------------------------------------------------

TIMEZONE=Europe/Berlin

DB_ENGINE=django.db.backends.postgresql
POSTGRES_HOST=db_recipes
POSTGRES_PORT=5432
POSTGRES_USER=djangouser
# ---------------------------- REQUIRED -------------------------
POSTGRES_PASSWORD=...
# ---------------------------------------------------------------
POSTGRES_DB=djangodb

FRACTION_PREF_DEFAULT=0
COMMENT_PREF_DEFAULT=1
SHOPPING_MIN_AUTOSYNC_INTERVAL=5
GUNICORN_MEDIA=0
REVERSE_PROXY_AUTH=0

```

### `docker-compose.yml`
When running with docker compose please provide your `docker-compose.yml`
```
version: "3"
services:
  db_recipes:
    restart: always
    image: postgres:11-alpine
    volumes:
      - ./postgresql:/var/lib/postgresql/data
    env_file:
      - ./.env

  web_recipes:
    image: vabene1111/recipes
    restart: always
    env_file:
      - ./.env
    volumes:
      - staticfiles:/opt/recipes/staticfiles
      - nginx_config:/opt/recipes/nginx/conf.d
      - ./mediafiles:/opt/recipes/mediafiles
    depends_on:
      - db_recipes

  nginx_recipes:
    image: nginx:mainline-alpine
    restart: always
    ports:
      - 23080:80
    env_file:
      - ./.env
    depends_on:
      - web_recipes
    volumes:
      - nginx_config:/etc/nginx/conf.d:ro
      - staticfiles:/static
      - ./mediafiles:/media

volumes:
  nginx_config:
  staticfiles:
```

### Logs
The following log entries show what nginx is logging for various attemps using an URL for import or using JSON directly:
```
2022/01/08 06:45:02 [error] 28#28: *403 upstream prematurely closed connection while reading response header from upstream, client: 172.18.0.1, server: localhost, request: "POST /api/recipe-from-source/ HTTP/1.0", upstream: "http://172.18.0.3:8080/api/recipe-from-source/", host: "ANONYMIZED.de", referrer: "https://ANONYMIZED.de/data/import/url"
2022/01/08 06:47:03 [error] 28#28: *405 upstream prematurely closed connection while reading response header from upstream, client: 172.18.0.1, server: localhost, request: "POST /api/recipe-from-source/ HTTP/1.0", upstream: "http://172.18.0.3:8080/api/recipe-from-source/", host: "ANONYMIZED.de", referrer: "https://ANONYMIZED.de/data/import/url"
2022/01/08 06:47:35 [error] 28#28: *407 upstream prematurely closed connection while reading response header from upstream, client: 172.18.0.1, server: localhost, request: "POST /api/recipe-from-source/ HTTP/1.0", upstream: "http://172.18.0.3:8080/api/recipe-from-source/", host: "ANONYMIZED.de", referrer: "https://ANONYMIZED.de/data/import/url"
2022/01/08 06:53:32 [error] 28#28: *478 upstream prematurely closed connection while reading response header from upstream, client: 172.18.0.1, server: localhost, request: "POST /api/recipe-from-source/ HTTP/1.0", upstream: "http://172.18.0.3:8080/api/recipe-from-source/", host: "ANONYMIZED.de", referrer: "https://ANONYMIZED.de/data/import/url"
2022/01/08 06:57:16 [error] 28#28: *519 upstream prematurely closed connection while reading response header from upstream, client: 172.18.0.1, server: localhost, request: "POST /data/import/url HTTP/1.0", upstream: "http://172.18.0.3:8080/data/import/url", host: "ANONYMIZED.de", referrer: "https://ANONYMIZED.de/data/import/url"
2022/01/08 07:02:03 [error] 28#28: *662 upstream prematurely closed connection while reading response header from upstream, client: 172.18.0.1, server: localhost, request: "POST /data/import/url HTTP/1.0", upstream: "http://172.18.0.3:8080/data/import/url", host: "ANONYMIZED.de", referrer: "https://ANONYMIZED.de/data/import/url"
2022/01/08 07:08:28 [error] 28#28: *678 upstream prematurely closed connection while reading response header from upstream, client: 172.18.0.1, server: localhost, request: "POST /data/import/url HTTP/1.0", upstream: "http://172.18.0.3:8080/data/import/url", host: "ANONYMIZED.de", referrer: "https://ANONYMIZED.de/data/import/url"
```

**Comment by smilerz** [comment]
You are using 2 nginx proxies?

**Comment by tdittmar** [comment]
Well, I'm running this on a vserver that is already able to route nginx to a docker container (this is configred in Plesk on the server). 

In my first attempt I left out the nginx container from the docker_compose.yml and directed that nginx instance directly to web_recpies. That seemed to work at first, but I was not able to show any media files (as stated in the docs).

So I decided to to use the nginx container (because then media files are served correctly), but then I don't seem to be able to import any recipes.

**Comment by smilerz** [comment]
the single nginx instance will work - but you need to mount & serve the media folder from the nginx proxy.  

I guess 2 proxies could work - but the failure you are experiencing is occurring in the interaction between those 2 proxies, I'm not sure we can help you.

**Comment by smilerz** [comment]
If you go to the System view - there is a DEBUG window that shows what headers Django is receiving.  If you compare the output between 1 proxy and 2 proxies that might point you in the direction of the problem.

**Comment by tdittmar** [comment]
> the single nginx instance will work - but you need to mount & serve the media folder from the nginx proxy.
> 
> I guess 2 proxies could work - but the failure you are experiencing is occurring in the interaction between those 2 proxies, I'm not sure we can help you.

I guess I'll try to make the single-nginx solution work.

**Comment by tdittmar** [comment]
OK. I got it to work so that media files are server properly using only my server's nginx proxying directly to the web_recipes container without using the nginx container.

However, I still cannot import recipes via JSON. The docker log for the web_recipes container shows the following:

`[2022-01-11 20:18:18 +0000] [1] [CRITICAL] WORKER TIMEOUT (pid:14)`

and I still get a 499 error from nginx. Any ideas?

**Comment by smilerz** [comment]
are you still having this issue?  if so, please provide DEBUG logs from the web container as well as browser console output

**Comment by smilerz** [comment]
closed due to lack of info

**Comment by mavaa** [comment]
@smilerz I might have the same, or a similar issue to this, but I'm not running two proxies. Trying to import this recipe, which seems fine when pasting the link: https://www.godt.no/oppskrift/8065/mettende-kyllingsalat

However when i click the "Import" button at the bottom of the page, it seems to hang and run into some sort of timeout.

I have been able to do this for other recipes earlier, but it seems to have stopped working recently (perhaps related to a recent upgrade).

Browser Request:

```
POST /data/import/url HTTP/1.1
Host: my-tandoor-host
Connection: keep-alive
Content-Length: 4534
Accept: application/json, text/plain, */*
X-CSRFToken: WxqcNie0Ih7BqDdYaTstIIWhXrKKubpQIMQCtpzvT8CzUAkxPANUkyJQrewyvxmP
X-Requested-With: XMLHttpRequest
User-Agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) QtWebEngine/5.15.2 Chrome/87.0.4280.144 Safari/537.36
Content-Type: application/json;charset=UTF-8
Accept-Language: en-US,en;q=0.9
DNT: 1
Origin: http://my-tandoor-host
Referer: http://my-tandoor-host/data/import/url
Accept-Encoding: gzip, deflate
Cookie: csrftoken=REMOVED; sessionid=REMOVED; shopping_entry_mode_simple=true; mealplan_settings=%7B%22displayPeriodUom%22%3A%22week%22%2C%22displayPeriodCount%22%3A2%2C%22startingDayOfWeek%22%3A1%2C%22displayWeekNumbers%22%3Atrue%7D
```

Browser response:

```
HTTP/1.1 502 Bad Gateway
Server: nginx/1.21.5
Date: Sun, 13 Feb 2022 17:39:06 GMT
Content-Type: text/html
Content-Length: 559
Connection: keep-alive
```

Server log for the request:

```
tandoor-web_recipes-1    | [2022-02-13 17:39:06 +0000] [1] [CRITICAL] WORKER TIMEOUT (pid:14)
tandoor-nginx_recipes-1  | 2022/02/13 17:39:06 [error] 22#22: *38 upstream prematurely closed connection while reading response header from upstream, client: clientIpHere, server: localhost, request: "POST /data/import/url HTTP/1.1", upstream: "http://server_ip:8080/data/import/url", host: "my-tandoor-host", referrer: "http://my-tandoor-host/data/import/url"
tandoor-web_recipes-1    | [2022-02-13 18:39:06 +0100] [14] [INFO] Worker exiting (pid: 14)
tandoor-nginx_recipes-1  | clientIpHere - - [13/Feb/2022:17:39:06 +0000] "POST /data/import/url HTTP/1.1" 502 559 "http://my-tandoor-host/data/import/url" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) QtWebEngine/5.15.2 Chrome/87.0.4280.144 Safari/537.36" "-"
tandoor-web_recipes-1    | [2022-02-13 17:39:06 +0000] [15] [INFO] Booting worker with pid: 15
```

Server info:
```
Gunicoren Media: False
Sqlite: False
Debug: False

SERVER_PROTOCOL:HTTP/1.0
REMOTE_ADDR:REMOVED
SERVER_PORT:8080

HTTP_HOST:my-tandoor-host
HTTP_CONNECTION:close
HTTP_UPGRADE_INSECURE_REQUESTS:1
HTTP_USER_AGENT:Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) QtWebEngine/5.15.2 Chrome/87.0.4280.144 Safari/537.36
HTTP_ACCEPT:text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9
HTTP_ACCEPT_LANGUAGE:en-US,en;q=0.9
HTTP_DNT:1
HTTP_REFERER:http://my-tandoor-host/search/
HTTP_ACCEPT_ENCODING:gzip, deflate
HTTP_COOKIE:csrftoken=REMOVED; sessionid=REMOVED; shopping_entry_mode_simple=true; mealplan_settings=%7B%22displayPeriodUom%22%3A%22week%22%2C%22displayPeriodCount%22%3A2%2C%22startingDayOfWeek%22%3A1%2C%22displayWeekNumbers%22%3Atrue%7D

wsgi.errors:<gunicorn.http.wsgi.WSGIErrorsWrapper object at 0x7fa8a9b69ee0>
wsgi.version:(1, 0)
wsgi.multithread:False
wsgi.multiprocess:False
wsgi.run_once:False
wsgi.file_wrapper:
wsgi.input_terminated:True
wsgi.input:<gunicorn.http.body.Body object at 0x7fa8a9b69850>
wsgi.url_scheme:http
```
```
Current Version: 1.0.4
Ref: af9a2a89ecc7e9a175cc35f371f5714befd8a4a6
```

**Comment by mavaa** [comment]
I just re-tried another recipe, which worked fine, so it's probably related to some specific recipes. This one worked: https://www.nrk.no/mat/gronnkal--og-granateplesalat-med-dijonvinaigrette-1.13370992

If relevant, here's the json that gets sent to the /data/import/url endpoint:
```
{
    "name": "Mettende kyllingsalat",
    "description": "En deilig gresk salat - med en vri! Denne serveres med hummus, grillet kylling, avokado og spirer. Lekkert og sunt.",
    "servings": 4,
    "prepTime": 25,
    "cookTime": 0,
    "image": "https://imbo.vgc.no/users/godt/images/75f75d506361eddc8971d77ddf5df5ba?t[]=thumbnail:width=1900,height=1425,fit=outbound&t[]=strip&t[]=compress:level=75&t[]=progressive&accessToken=f7d4390020a3aee66ceb44e9b39a45fd4c21596d87225c7d78d544861ad65183",
    "keywords": [
        {
            "id": 5965290,
            "text": "salat"
        },
        {
            "id": 8098419,
            "text": "middag"
        },
        {
            "id": 6998661,
            "text": "lunsj"
        }
    ],
    "recipeIngredient": [
        {
            "amount": 2,
            "unit": {
                "text": "stk",
                "id": 14792
            },
            "ingredient": {
                "text": "hjertesalat",
                "id": 68393
            },
            "note": "",
            "original": "2 stk hjertesalat"
        },
        {
            "amount": 0.5,
            "unit": {
                "text": "stk",
                "id": 40399
            },
            "ingredient": {
                "text": "agurk",
                "id": 35212
            },
            "note": "",
            "original": "0.5 stk agurk"
        },
        {
            "amount": 200,
            "unit": {
                "text": "g",
                "id": 80580
            },
            "ingredient": {
                "text": "tomat",
                "id": 98131
            },
            "note": "",
            "original": "200 g tomat"
        },
        {
            "amount": 150,
            "unit": {
                "text": "g",
                "id": 22553
            },
            "ingredient": {
                "text": "fetaost",
                "id": 96816
            },
            "note": "",
            "original": "150 g fetaost"
        },
        {
            "amount": 2,
            "unit": {
                "text": "dl",
                "id": 42970
            },
            "ingredient": {
                "text": "hummus",
                "id": 40337
            },
            "note": "",
            "original": "2 dl hummus"
        },
        {
            "amount": 150,
            "unit": {
                "text": "g",
                "id": 41901
            },
            "ingredient": {
                "text": "kikerter",
                "id": 12829
            },
            "note": "",
            "original": "150 g kikerter"
        },
        {
            "amount": 1,
            "unit": {
                "text": "stk",
                "id": 65354
            },
            "ingredient": {
                "text": "avokado",
                "id": 99061
            },
            "note": "",
            "original": "1 stk avokado"
        },
        {
            "amount": 2,
            "unit": {
                "text": "stk",
                "id": 15894
            },
            "ingredient": {
                "text": "kyllingbryst",
                "id": 13157
            },
            "note": "",
            "original": "2 stk kyllingbryst"
        },
        {
            "amount": 0,
            "unit": {
                "text": "",
                "id": 24107
            },
            "ingredient": {
                "text": "spirer",
                "id": 17896
            },
            "note": "",
            "original": "spirer"
        },
        {
            "amount": 0,
            "unit": {
                "text": "",
                "id": 63193
            },
            "ingredient": {
                "text": "friske urter",
                "id": 58884
            },
            "note": "",
            "original": "friske urter"
        },
        {
            "amount": 0,
            "unit": {
                "text": "",
                "id": 32974
            },
            "ingredient": {
                "text": "chiliflak",
                "id": 91596
            },
            "note": "",
            "original": "chiliflak"
        },
        {
            "amount": 0,
            "unit": {
                "text": "",
                "id": 90666
            },
            "ingredient": {
                "text": "syltet r\u00f8dl\u00f8k",
                "id": 54140
            },
            "note": "",
            "original": "syltet r\u00f8dl\u00f8k"
        },
        {
            "amount": 0,
            "unit": {
                "text": "",
                "id": 40920
            },
            "ingredient": {
                "text": "oliven",
                "id": 87983
            },
            "note": "",
            "original": "oliven"
        },
        {
            "amount": 0,
            "unit": {
                "text": "",
                "id": 18843
            },
            "ingredient": {
                "text": "olivenolje",
                "id": 25153
            },
            "note": "",
            "original": "olivenolje"
        },
        {
            "amount": 0,
            "unit": {
                "text": "",
                "id": 31868
            },
            "ingredient": {
                "text": "r\u00f8dvinseddik",
                "id": 40987
            },
            "note": "",
            "original": "r\u00f8dvinseddik"
        },
        {
            "amount": 0,
            "unit": {
                "text": "",
                "id": 88750
            },
            "ingredient": {
                "text": "salt",
                "id": 50645
            },
            "note": "",
            "original": "salt"
        },
        {
            "amount": 0,
            "unit": {
                "text": "",
                "id": 23622
            },
            "ingredient": {
                "text": "pepper",
                "id": 79941
            },
            "note": "",
            "original": "pepper"
        },
        {
            "amount": 2,
            "unit": {
                "text": "stk",
                "id": 49452
            },
            "ingredient": {
                "text": "r\u00f8dl\u00f8k",
                "id": 11563
            },
            "note": "",
            "original": "2 stk r\u00f8dl\u00f8k"
        },
        {
            "amount": 1,
            "unit": {
                "text": "dl",
                "id": 69701
            },
            "ingredient": {
                "text": "eplesidereddik",
                "id": 33607
            },
            "note": "",
            "original": "1 dl eplesidereddik"
        },
        {
            "amount": 2,
            "unit": {
                "text": "dl",
                "id": 32035
            },
            "ingredient": {
                "text": "sukker",
                "id": 14641
            },
            "note": "",
            "original": "2 dl sukker"
        },
        {
            "amount": 2.5,
            "unit": {
                "text": "dl",
                "id": 36222
            },
            "ingredient": {
                "text": "vann",
                "id": 37524
            },
            "note": "",
            "original": "2.5 dl vann"
        },
        {
            "amount": 1,
            "unit": {
                "text": "klype",
                "id": 58720
            },
            "ingredient": {
                "text": "salt",
                "id": 45595
            },
            "note": "",
            "original": "1 klype salt"
        }
    ],
    "recipeInstructions": "Vask og rens salaten og skj\u00e6r den i grove strimler. Skj\u00e6r agurk og tomater i jevne sm\u00e5 biter. Rens avokadoen og skj\u00e6r den i terninger eller tynne skiver. Skj\u00e6r kyllingbrystene i tynne skiver eller terninger.  \nFordel salat, agurk og tomat i fire boller eller bokser. Drypp over litt god olivenolje og noen dr\u00e5per eddik. Krydre lett med salt og kvernet pepper. Topp med smuldret fetaost, hummus, kikerter, avokado og kylling. Avslutt med spirer, urter, oliven og syltet r\u00f8dl\u00f8k. Drypp over litt ekstra olje og en liten klype chiliflak.  \nSyltet r\u00f8dl\u00f8k: Rens og skj\u00e6r l\u00f8ken i tynne skiver. Ha l\u00f8ken over p\u00e5 et passende sylteglass. Kok opp eddik, sukker og vann i en kasserolle. La det sm\u00e5koke til alt sukkeret er oppl\u00f8st. Tilsett en god klype salt og hell den varme syltelaken over l\u00f8ken (l\u00f8ken skal v\u00e6re dekket av lake). Sett p\u00e5 lokket og la l\u00f8ken st\u00e5 i romtemperatur til avkj\u00f8ling. L\u00f8ken kan brukes straks den er avkj\u00f8lt, men blir enda bedre hvis den kan f\u00e5 st\u00e5 natten over. Oppbevares kj\u00f8lig.\n\nImported from https://www.godt.no/oppskrift/8065/mettende-kyllingsalat",
    "url": "https://www.godt.no/oppskrift/8065/mettende-kyllingsalat",
    "all_keywords": false
}
```

I see there are some regional charackters, æøå, maybe that's a problem (Or i've done a misconfiguration)?

**Comment by mavaa** [comment]
Tried with json from my previous comment in the source tab, but it doesn't seem to recognize any of the fields. I can go over in manual mode and add entries manually, but for ingredients I have to add unit, amount, name etc manually for each entry, so it's not optimal.

But I have a hard time understanding how it can be a timeout problem for the automatic mode. When I paste the URL all the fields get filled out correctly:

![2022_02_14_22_04_24](https://user-images.githubusercontent.com/1224973/153946328-a8c57946-8af8-4626-9905-201c836ad97c.png)

I also tried removing the image and instructions in case they were causing trouble.

It's when hitting the final "Import" button at the bottom I get an error:
![2022_02_14_22_05_37](https://user-images.githubusercontent.com/1224973/153946510-5821f31a-ed32-42a0-83f2-88025c1d49a1.png)
![2022_02_14_22_06_05](https://user-images.githubusercontent.com/1224973/153946529-9a525eea-3029-4b16-be92-1f7beec6390b.png)

Maybe related, I found this warning in the console:

![2022_02_14_22_07_17](https://user-images.githubusercontent.com/1224973/153946591-808b2221-f7d5-471d-bdaf-9b6cfc5e5c8a.png)

Google tells me it might be because of running the site without SSL, but I'm just running it on a local closed-off network so I haven't bothered yet. Could that be what's causing the issue?
